src/model/ScheduleDay.cpp

```diff
@@ -176,42 +181,123 @@ namespace model {
         return 0.0;
       }
 
-      std::vector<double> values = this->values();          // these are already sorted
-      std::vector<openstudio::Time> times = this->times();  // these are already sorted
+      TimeSeries ts = this->timeSeries();
 
-      unsigned N = times.size();
+      DateTimeVector dateTimes = ts.dateTimes();
+      Vector values = ts.values();
+
+      unsigned N = dateTimes.size();
       OS_ASSERT(values.size() == N);
 
       if (N == 0) {
         return 0.0;
       }
 
-      openstudio::Vector x(N + 2);
-      openstudio::Vector y(N + 2);
-
-      x[0] = -0.000001;
-      y[0] = 0.0;
+      Vector x(N + 2);
+      Vector y(N + 2);
+
+      x[0] = 0.0;
+      std::string interpolatetoTimestep = this->interpolatetoTimestep();
+      if (istringEqual("No", interpolatetoTimestep)) {
+        y[0] = values[0];
+      } else if (istringEqual("Average", interpolatetoTimestep)) {
+        y[0] = values[0];
+      } else if (istringEqual("Linear", interpolatetoTimestep)) {
+        y[0] = 0.0;
+      }
 
       for (unsigned i = 0; i < N; ++i) {
-        x[i + 1] = times[i].totalDays();
+        openstudio::Time t = dateTimes[i].time();
+        if (t.totalDays() == 0.0) {  // this is 00:00:00 from the next day
+          t = openstudio::Time(0, 24, 0);
+        }
+
+        x[i + 1] = t.totalDays();
         y[i + 1] = values[i];
       }
 
-      x[N + 1] = 1.000001;
-      y[N + 1] = 0.0;
+      x[N + 1] = 1.0;
+      y[N + 1] = values[N - 1];
 
-      InterpMethod interpMethod;
-      if (this->interpolatetoTimestep()) {
-        interpMethod = LinearInterp;
-      } else {
-        interpMethod = HoldNextInterp;
-      }
-
-      double result = interp(x, y, time.totalDays(), interpMethod, NoneExtrap);
+      double result = interp(x, y, time.totalDays(), HoldNextInterp, NoneExtrap);
```

Comment:
I really feel like you ever need to pass two values to that call...

As I wrote on https://github.com/NREL/OpenStudio/issues/5001#issuecomment-1998619932

> doesn't the getValues routine look inefficient? If I'm understanding correctly, you only need to find the time before and the time after and the corresponding, you can break early. I know, this is a scheduleday, so we're talking about 60 (timestep max in E+ If I recall correctly) * 24 hours datapoints max, so it's not noticeable, but still


And you should probably just two that on values & times, not on timeSeries (which is interpolated already, and contains the entire day)